network/src/gossip.rs

```diff
@@ -129,57 +207,586 @@ pub(crate) struct MessageValidationData {
 }
 
 impl MessageValidationData {
-	fn check_statement(&self, relay_parent: &Hash, statement: &SignedStatement) -> bool {
+	fn check_statement(&self, relay_parent: &Hash, statement: &SignedStatement) -> Result<(), ()> {
 		let sender = match self.index_mapping.get(&statement.sender) {
 			Some(val) => val,
-			None => return false,
+			None => return Err(()),
 		};
 
-		self.authorities.contains(&sender) &&
+		let good = self.authorities.contains(&sender) &&
 			::polkadot_validation::check_statement(
 				&statement.statement,
 				&statement.signature,
 				sender.clone(),
 				relay_parent,
-			)
+			);
+
+		if good {
+			Ok(())
+		} else {
+			Err(())
+		}
+	}
+}
+
+// knowledge about attestations on a single parent-hash.
+#[derive(Default)]
+struct Knowledge {
+	candidates: HashSet<Hash>,
+}
+
+impl Knowledge {
+	// whether the peer is aware of a candidate with given hash.
+	fn is_aware_of(&self, candidate_hash: &Hash) -> bool {
+		self.candidates.contains(candidate_hash)
+	}
+
+	// note that the peer is aware of a candidate with given hash.
+	fn note_aware(&mut self, candidate_hash: Hash) {
+		self.candidates.insert(candidate_hash);
+	}
+}
+
+struct PeerData {
+	live: HashMap<Hash, Knowledge>,
+}
+
+impl PeerData {
+	fn knowledge_at_mut(&mut self, parent_hash: &Hash) -> Option<&mut Knowledge> {
+		self.live.get_mut(parent_hash)
+	}
+}
+
+struct OurView {
+	live_sessions: Vec<(Hash, SessionView)>,
```

Comment:
I assume that all chain heads have a different `Hash` anyway, so probably this should also be a `HashMap` but with some size assertions?

Reply:
it's way more lean to use a `Vec` for the small number of items.